src/spyglass/utils/dj_graph.py

```diff
@@ -609,7 +697,7 @@ def add_leaves(
 
     # ------------------------------ Graph Traversal --------------------------
 
-    def cascade(self, show_progress=None, direction="up") -> None:
+    def cascade(self, show_progress=None, direction="up", warn=True) -> None:
```

Comment:
Usage issue: 
When trying deletes in the lab database, I'm getting errors where custom user tables that are downstream of the table I'm deleting from block deletion because they are not loaded into the graph.  This is true even if the entry I'm deleting does not apply to those tables.  Is there an appropriate place to restrict in cascade to prevent this?

ex)
```python
from spyglass.common import Session
from spyglass.position.v1 import TrodesPosV1
from spyglass.decoding import PositionGroup

key = {"nwb_file_name":"minirec20230622_.nwb"}
(TrodesPosV1 & key).delete_downstream_parts(dry_run=True)
```
raises:
```python
{
	"name": "ValueError",
	"message": "Table `jguidera_head_speed`.`__epoch_low_head_speed_periods__upstream` not found in graph.
\tPlease import this table and rerun",
	"stack": "---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Untitled-1.ipynb Cell 1 line 7
      <a href='vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W0sdW50aXRsZWQ%3D?line=4'>5</a> key = {\"nwb_file_name\":\"minirec20230622_.nwb\"}
      <a href='vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W0sdW50aXRsZWQ%3D?line=5'>6</a> key = {\"nwb_file_name\":\"Winnie20220712_.nwb\"}
----> <a href='vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W0sdW50aXRsZWQ%3D?line=6'>7</a> (TrodesPosV1 & key).delete_downstream_parts(dry_run=True)

File ~/Documents/frank_lab/code/spyglass/src/spyglass/utils/dj_mixin.py:380, in SpyglassMixin.delete_downstream_parts(self, restriction, dry_run, reload_cache, disable_warning, return_graph, **kwargs)
    376     _ = self.__dict__.pop(\"_part_masters\", None)
    378 restriction = restriction or self.restriction or True
--> 380 restr_graph = RestrGraph(
    381     seed_table=self,
    382     leaves={self.full_table_name: restriction},
    383     direction=\"down\",
    384     cascade=True,
    385     verbose=False,
    386 )
    388 if return_graph:
    389     return restr_graph

File ~/Documents/frank_lab/code/spyglass/src/spyglass/utils/dj_graph.py:579, in RestrGraph.__init__(self, seed_table, leaves, destinations, direction, cascade, verbose, **kwargs)
    577 for dir in dir_list:
    578     self._log_truncate(f\"Start {dir:<4} : {self.leaves}\")
--> 579     self.cascade(direction=dir)
    580     self.cascaded = False
    581     self.visited -= self.leaves

File ~/Documents/frank_lab/code/spyglass/src/spyglass/utils/dj_graph.py:725, in RestrGraph.cascade(self, show_progress, direction, warn)
    721     restr = self._get_restr(table)
    722     self._log_truncate(
    723         f\"Start  {direction:<4}: {self._camel(table)}, {restr}\"
    724     )
--> 725     self.cascade1(table, restr, direction=direction)
    727 self.cascaded = True  # Mark here so next step can use `restr_ft`
    728 self.cascade_files()

File ~/Documents/frank_lab/code/spyglass/src/spyglass/utils/dj_graph.py:440, in AbstractGraph.cascade1(self, table, restriction, direction, replace, count, **kwargs)
    437 if next_restr == [\"False\"]:  # Stop cascade if empty restriction
    438     continue
--> 440 self.cascade1(
    441     table=next_table,
    442     restriction=next_restr,
    443     direction=direction,
    444     replace=replace,
    445     count=count + 1,
    446 )

File ~/Documents/frank_lab/code/spyglass/src/spyglass/utils/dj_graph.py:440, in AbstractGraph.cascade1(self, table, restriction, direction, replace, count, **kwargs)
    437 if next_restr == [\"False\"]:  # Stop cascade if empty restriction
    438     continue
--> 440 self.cascade1(
    441     table=next_table,
    442     restriction=next_restr,
    443     direction=direction,
    444     replace=replace,
    445     count=count + 1,
    446 )

File ~/Documents/frank_lab/code/spyglass/src/spyglass/utils/dj_graph.py:430, in AbstractGraph.cascade1(self, table, restriction, direction, replace, count, **kwargs)
    427     self._log_truncate(f\"{reason}: {self._camel(next_table)}\")
    428     continue
--> 430 next_restr = self._bridge_restr(
    431     table1=table,
    432     table2=next_table,
    433     restr=restriction,
    434     **data,
    435 )
    437 if next_restr == [\"False\"]:  # Stop cascade if empty restriction
    438     continue

File ~/Documents/frank_lab/code/spyglass/src/spyglass/utils/dj_graph.py:308, in AbstractGraph._bridge_restr(self, table1, table2, restr, direction, attr_map, aliased, **kwargs)
    305     attr_map = edge.get(\"attr_map\")
    307 ft1 = self._get_ft(table1) & restr
--> 308 ft2 = self._get_ft(table2)
    310 if len(ft1) == 0:
    311     return [\"False\"]

File ~/Documents/frank_lab/code/spyglass/src/spyglass/utils/dj_graph.py:257, in AbstractGraph._get_ft(self, table, with_restr, warn)
    254 else:
    255     restr = True
--> 257 if not (ft := self._get_node(table).get(\"ft\")):
    258     ft = FreeTable(self.connection, table)
    259     self._set_node(table, \"ft\", ft)

File ~/Documents/frank_lab/code/spyglass/src/spyglass/utils/dj_graph.py:178, in AbstractGraph._get_node(self, table)
    176 table = self._ensure_names(table)
    177 if not (node := self.graph.nodes.get(table)):
--> 178     raise ValueError(
    179         f\"Table {table} not found in graph.\"
    180         + \"\
\\tPlease import this table and rerun\"
    181     )
    182 return node

ValueError: Table `jguidera_head_speed`.`__epoch_low_head_speed_periods__upstream` not found in graph.
\tPlease import this table and rerun"
}
```

Reply:
I think I could stop cascading if the path went outside of common prefixes, but that would prevent use of this tool for deleting from one's custom schemas. Should this `try/except` handle things differently for outside prefixes? Only throwing a warning for custom tables?  If so, the user would hit the part delete error later and may not know to import their custom table and retry

I think I could hit this error and try the current restriction on a FreeTable and ignore it if it looks like a dead end - but that would mean some false negative dead ends if the table had aliases. The reason behind enforcing the import was to get that graph data that isn't available from the FreeTable

Reply:
I added a commit to warn of an ignored table and treat it as a dead end

Reply:
To check my understanding; 
- This will prevent cascading and `delete_downstream_parts` from propagating down to a hypothetical custom merge table not loaded in the namespace. 
- loading all tables on the path to the hypothetical merge table will allow it to work
- custom tables that don't need `delete_downstream_parts` will still be deleted by the super().delete call(s) that happen

Reply:
Also, I'm hitting a new issue.  

When Running:
```python
from spyglass.common import Session
from spyglass.position.v1 import TrodesPosV1
from spyglass.decoding import PositionGroup

key = {"nwb_file_name":"minirec20230622_.nwb"}
(TrodesPosV1 & key).delete_downstream_parts(dry_run=True)
```

I hit namespace errors at tables downstream of merge tables (`RippleTimesV1`, `LinearizedPositionV1`, `MuaEventsV1`) even if they are not populated for the key.
<details>

```python
{
	"name": "ValueError",
	"message": "Table `ripple_v1`.`__ripple_times_v1` not found in graph.
\tPlease import this table and rerun",
	"stack": "---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[1], line 10
      5 # from spyglass.ripple.v1 import RippleTimesV1
      6 # from spyglass.linearization.v1 import LinearizedPositionV1
      7 # from spyglass.mua.v1 import MuaEventsV1
      9 key = {\"nwb_file_name\":\"minirec20230622_.nwb\"}
---> 10 (TrodesPosV1 & key).delete_downstream_parts(dry_run=True)

File ~/Documents/spyglass/src/spyglass/utils/dj_mixin.py:395, in SpyglassMixin.delete_downstream_parts(self, restriction, dry_run, reload_cache, disable_warning, return_graph, verbose, **kwargs)
    391     _ = self.__dict__.pop(\"_part_masters\", None)
    393 restriction = restriction or self.restriction or True
--> 395 restr_graph = RestrGraph(
    396     seed_table=self,
    397     leaves={self.full_table_name: restriction},
    398     direction=\"down\",
    399     cascade=True,
    400     verbose=verbose,
    401 )
    403 if return_graph:
    404     return restr_graph

File ~/Documents/spyglass/src/spyglass/utils/dj_graph.py:598, in RestrGraph.__init__(self, seed_table, leaves, destinations, direction, cascade, verbose, **kwargs)
    596 for dir in dir_list:
    597     self._log_truncate(f\"Start {dir:<4} : {self.leaves}\")
--> 598     self.cascade(direction=dir)
    599     self.cascaded = False
    600     self.visited -= self.leaves

File ~/Documents/spyglass/src/spyglass/utils/dj_graph.py:744, in RestrGraph.cascade(self, show_progress, direction, warn)
    740     restr = self._get_restr(table)
    741     self._log_truncate(
    742         f\"Start  {direction:<4}: {self._camel(table)}, {restr}\"
    743     )
--> 744     self.cascade1(table, restr, direction=direction)
    746 self.cascaded = True  # Mark here so next step can use `restr_ft`
    747 self.cascade_files()

File ~/Documents/spyglass/src/spyglass/utils/dj_graph.py:455, in AbstractGraph.cascade1(self, table, restriction, direction, replace, count, **kwargs)
    452 if next_restr == [\"False\"]:  # Stop cascade if empty restriction
    453     continue
--> 455 self.cascade1(
    456     table=next_table,
    457     restriction=next_restr,
    458     direction=direction,
    459     replace=replace,
    460     count=count + 1,
    461 )

File ~/Documents/spyglass/src/spyglass/utils/dj_graph.py:455, in AbstractGraph.cascade1(self, table, restriction, direction, replace, count, **kwargs)
    452 if next_restr == [\"False\"]:  # Stop cascade if empty restriction
    453     continue
--> 455 self.cascade1(
    456     table=next_table,
    457     restriction=next_restr,
    458     direction=direction,
    459     replace=replace,
    460     count=count + 1,
    461 )

File ~/Documents/spyglass/src/spyglass/utils/dj_graph.py:445, in AbstractGraph.cascade1(self, table, restriction, direction, replace, count, **kwargs)
    442     self._log_truncate(f\"{reason}: {self._camel(next_table)}\")
    443     continue
--> 445 next_restr = self._bridge_restr(
    446     table1=table,
    447     table2=next_table,
    448     restr=restriction,
    449     **data,
    450 )
    452 if next_restr == [\"False\"]:  # Stop cascade if empty restriction
    453     continue

File ~/Documents/spyglass/src/spyglass/utils/dj_graph.py:323, in AbstractGraph._bridge_restr(self, table1, table2, restr, direction, attr_map, aliased, **kwargs)
    321 # May return empty table if outside imported and outside spyglass
    322 ft1 = self._get_ft(table1) & restr
--> 323 ft2 = self._get_ft(table2)
    325 if len(ft1) == 0 or len(ft2) == 0:
    326     return [\"False\"]

File ~/Documents/spyglass/src/spyglass/utils/dj_graph.py:258, in AbstractGraph._get_ft(self, table, with_restr, warn)
    255 else:
    256     restr = True
--> 258 if not (ft := self._get_node(table).get(\"ft\")):
    259     ft = FreeTable(self.connection, table)
    260     self._set_node(table, \"ft\", ft)

File ~/Documents/spyglass/src/spyglass/utils/dj_graph.py:179, in AbstractGraph._get_node(self, table)
    177 table = self._ensure_names(table)
    178 if not (node := self.graph.nodes.get(table)):
--> 179     raise ValueError(
    180         f\"Table {table} not found in graph.\"
    181         + \"\
\\tPlease import this table and rerun\"
    182     )
    183 return node

ValueError: Table `ripple_v1`.`__ripple_times_v1` not found in graph.
\tPlease import this table and rerun"
}
```
</details>

To get it to run properly I needed the following imports
```python
from spyglass.common import Session
from spyglass.position.v1 import TrodesPosV1

from spyglass.ripple.v1 import RippleTimesV1
from spyglass.linearization.v1 import LinearizedPositionV1
from spyglass.mua.v1 import MuaEventsV1

key = {"nwb_file_name":"minirec20230622_.nwb"}
(TrodesPosV1 & key).delete_downstream_parts(dry_run=True)
```